### ISSUE TITLE
issues involving variable initialisation

### ISSUE THREAD

**Issue by Neospiro** [colobot/colobot#872]
Hi,

I'm working a lot with the new features, but I have some problems : if there are ways to know if objets or ints are not initialized, that's not the case for points and bools.

Example : (will compile but not work) (if it don't compile, I will correct this when I could)
```
public extern object::bug(){
  base mybase(); 
}

class base{
  point center;

  base(){
    //if(this.center=null) dont pass compilation
    goto(this.center);
  }
}
```

Here is how I have to fix it : 

```
public extern object::bug(){
  base mybase(this); 
}

class base{
  point center;
  bool centerInit=false;

  base(object o){
    if(centerInit!=false){
      goto(this.center);  
    } else {
      this.center=o.position;
    }
    
  }

}
```

**Comment by krzys-h** [comment]
I don't think not being able to see if variable was already initialized or not is a problem since it's how it works in many languages - you usually get garbage data that just happened to be lying in memory (C++) or either some sane default values or compile error (Java). The `== null` syntax can be used only with classes (or `object` type as it is also a class), not with ordinary variables. `point` is special here as while being a class it's usually treated like normal variable types (in source code it's called an 'intrinsic class')

The expected behavior in the first example would be to fail with a runtime error. It does that with base types like `int` and that's how I remember it always working, but it seems like you are right that it does not do that with `point` and just initializes the variable with x=0 y=0 z=0. I'm not sure whether it always worked like that or if recent changes broke something, and even if it did maybe it would be a good idea to fix it. It may be because it's an intrinsic type. I'll wait on opinion from others about that (@colobot/active-members @melex750 @CyborgMaster)

As for returning nothing from a function, it's a long known limitation of CBOT - it doesn't check whether you actually returned the expected value from function or not. You are not supposed to be able to return without a value in non-void function, and it should cause a compile (or at least a runtime) error. Not returning a correct value usually just crashes the game. See #30 (wow, that's one of the first issues on this repo :open_mouth:)

**Comment by Neospiro** [comment]
I work a lot on high level environments so I don't really know low levels behaviours :/
But, in colobot, int have the "nan", and many function returns "nan" when it should return an integer. Others return null when it should return an object like "radar" or "search".
and the "space" function should return null, and return the center when it fail...

**Comment by melex750** [comment]
Arrays and 'point' are the only variables that can be passed as an argument
uninitialized, everything else throws runtime error "Variable not initialized".

Checking if a variable is initialized happens here:
https://github.com/colobot/colobot/blob/dev/src/CBot/CBotInstr/CBotExprVar.cpp#L248
https://github.com/colobot/colobot/blob/dev/src/CBot/CBotVar/CBotVar.h#L264

Point always reports initialized because of this:
https://github.com/colobot/colobot/blob/dev/src/CBot/CBotVar/CBotVar.cpp#L342

This is because assigning values to it's members will
not mark the point itself as initialized, example:
```c++
point p;
p.x = 1.1;
p.y = 2.2;
p.z = 3.3;
p.point(1,2,3); // even calling the constructor
message( "" + p ); // here 'p' is not initialized
```
You can set point to an invalid state like this:
```c++
p.x = nan;
p.y = nan;
p.z = nan;
if (p.x == nan) /*do stuff*/; // this works
```
We could make "x,y,and z" init with```nan```by default
and make point constructor accept```nan```as arguments and
no-args point constructor could set each component to```nan```.

Maybe built-in functions could throw an error if x or y are not initialized
or if x or y ==```nan```.

**Comment by krzys-h** [comment]
@melex750 Oh, that makes sense now. I think your proposed solution should work nicely too, and it shouldn't affect any existing programs (and if it does - well, technically these programs were using an undefined behavior so... ;))

**Comment by krzys-h** [comment]
And by the way, `p.point(1, 2, 3)` is a really strange syntax, does something like that actually work in any other programming language?

**Comment by tomaszkax86** [comment]
Directly calling constructor on existing object is not a usual feature, but it can be useful. C++ has something similar.

**Comment by hexagonrecursion** [comment]
Can someone clarify what the issue is?

1. It is impossible to detect whether a point was initialized - wontfix; not a bug
2. `bool isBug(){ return; }` now produces a compile error - as it should
3. `bool isBug(){  }` now produces a compile error - as it should

**Comment by melex750** [comment]
>> Can someone clarify what the issue is?

4. Maybe it was kept for the idea of initializing point by default to {nan, nan, nan},

The members of 'point' are currently assigned zero by default which is ok. At some point x,y,z were not initialized at all.

**Comment by hexagonrecursion** [comment]
In that case the title should be "Discussion: should point default to nan, nan, nan instead of 0, 0, 0?"
